nntrainer/layers/lstmcell_core.cpp

```diff
@@ -0,0 +1,519 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 hyeonseok lee <hs89.lee@samsung.com>
+ *
+ * @file   lstmcell_core.cpp
+ * @date   25 November 2021
+ * @brief  This is LSTMCellCore Layer Class of Neural Network
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author hyeonseok lee <hs89.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include <cmath>
+#include <lazy_tensor.h>
+#include <lstmcell_core.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+// Todo: remove this if sharing wt_idx with other lstm variant is enabled
+#define ENABLE_SHARING_WT_IDX 0
+
+namespace nntrainer {
+
+void fillLayerInitContext(InitLayerContext &context,
+                          const InitLayerContext &core_context) {
+  /** real set the input flags */
+  auto const &input_dims = context.getInputDimensions();
+  for (unsigned int idx = 0; idx < core_context.getNumInputs(); idx++) {
+    context.setDynDimFlagInputDimension(idx, input_dims[idx].getDynDimFlag());
+    context.setEffDimFlagInputDimension(idx, input_dims[idx].getEffDimFlag());
+  }
+
+  /** real request of tensors */
+  for (auto const &ts : core_context.getTensorsSpec())
+    context.requestTensor(ts);
+
+  /** real request of weights */
+  for (auto const &ws : core_context.getWeightsSpec())
+    context.requestWeight(ws);
+}
+
+void fillWeights(std::vector<Weight> &weights, const RunLayerContext &context,
+                 bool training, const unsigned int max_timestep,
+                 const unsigned int timestep, bool test) {
+  weights.resize(context.getNumWeights());
+  for (unsigned int i = 0; i < context.getNumWeights(); ++i) {
+    if (training && (!test || i < 3u)) {
+      weights[i] = Weight(context.getWeight(i), context.getWeightGrad(i),
+                          context.getWeightName(i));
+    } else {
+      weights[i] =
+        Weight(context.getWeight(i), Tensor(), context.getWeightName(i));
+    }
+  }
+}
+
+const std::vector<Weight *> getWeights(std::vector<Weight> &weights) {
+  std::vector<Weight *> ret(weights.size());
+  for (unsigned int i = 0; i < weights.size(); ++i) {
+    ret[i] = &weights[i];
+  }
+  return ret;
+}
+
+void fillInputs(std::vector<Var_Grad> &inputs, RunLayerContext &context,
```

Comment:
The way fillInputs, fillOutputs, fillDerivatives is designed that it reduces the logic burden out of the forwarding and backwrading. This definitely has the advantage of the clarification of the forwarding and backwarding core funcitons but come at the cost of possibly reduced optimizations and overhead of doing the work every iteration. 
This is good for now, but will need some benchmarking later.